File: java/db-password.java
/*
취약점 개요
JDBC 커넥션 객체를 생성할 때, DB 서버와 연결 시 사용하는 비밀번호 값은 하드코딩이나 빈 값으로 지정하면 안된다.

password 값을 System.getProperty() 함수를 이용하여 불러와서 사용한다

*/


// Noncompliant Code Example
Connection conn = DriverManager.getConnection("jdbc:derby:memory:myDB;create=true", "login", "");

// Compliant Solution
String password = System.getProperty("database.password");
Connection conn = DriverManager.getConnection("jdbc:derby:memory:myDB;create=true", "login", password);

File: python/db-password.py
'''
취약점 개요
DB 연결 시 null 비밀번호를 사용하거나, 하드코딩처리 해선 안된다.

- 환경변수에 저장된 값을 들고오는 방식으로 처리한다.

'''

# Noncompliant Code Example
# Flask-SQLAlchemy
def configure_app(app):
    app.config['SQLALCHEMY_DATABASE_URI'] = "postgresql://user:@domain.com" # Noncompliant

# Django - settings.py
DATABASES = {
    'postgresql_db': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': 'quickdb',
        'USER': 'sonarsource',
        'PASSWORD': '', # Noncompliant
        'HOST': 'localhost',
        'PORT': '5432'
    }
}

# mysql/mysql-connector-python
from mysql.connector import connection

connection.MySQLConnection(host='localhost', user='sonarsource', password='')  # Noncompliant


# Compliant Solution
# Flask-SQLAlchemy
def configure_app(app, pwd):
    app.config['SQLALCHEMY_DATABASE_URI'] = f"postgresql://user:{pwd}@domain.com" # Compliant

# Django - settings.py
import os

DATABASES = {
    'postgresql_db': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': 'quickdb',
        'USER': 'sonarsource',
        'PASSWORD': os.getenv('DB_PASSWORD'),      # Compliant
        'HOST': 'localhost',
        'PORT': '5432'
    }
}

# mysql/mysql-connector-python
from mysql.connector import connection
import os

db_password = os.getenv('DB_PASSWORD')
connection.MySQLConnection(host='localhost', user='sonarsource', password=db_password)  # Compliant